rskj-core/src/main/java/co/rsk/mine/ForkDetectionDataCalculator.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2019 RSK Labs Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.mine;
+
+import co.rsk.bitcoinj.core.Context;
+import co.rsk.bitcoinj.core.NetworkParameters;
+import co.rsk.bitcoinj.params.RegTestParams;
+import org.ethereum.core.Block;
+import org.ethereum.core.BlockHeader;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class ForkDetectionDataCalculator {
+
+    private static final int CPV_SIZE = 7;
+
+    private static final int CPV_JUMP_FACTOR = 64;
+
+    private static final int NUMBER_OF_UNCLES = 32;
+
+    // + 1 because genesis block can't be used since it does not contain a valid BTC header
+    private static final int MIN_MAINCHAIN_SIZE = CPV_SIZE * CPV_JUMP_FACTOR + 1;
+
+    public byte[] calculate(List<Block> mainchainBlocks) {
+        if (mainchainBlocks.size() < MIN_MAINCHAIN_SIZE) {
+            return new byte[0];
+        }
+
+        List<BlockHeader> mainchainBlockHeaders = mainchainBlocks
+                .stream()
+                .map(Block::getHeader)
+                .collect(Collectors.toList());
+
+        return calculateWithBlockHeaders(mainchainBlockHeaders);
+    }
+
+    public byte[] calculateWithBlockHeaders(List<BlockHeader> mainchainBlockHeaders) {
+        if (mainchainBlockHeaders.size() < MIN_MAINCHAIN_SIZE) {
+            return new byte[0];
+        }
+
+        byte[] forkDetectionData = new byte[12];
+
+        byte[] commitToParentsVector = buildCommitToParentsVector(mainchainBlockHeaders);
+        System.arraycopy(commitToParentsVector, 0, forkDetectionData, 0, 7);
+
+        short numberOfUncles = getNumberOfUncles(mainchainBlockHeaders);
+        forkDetectionData[7] = ByteBuffer.allocate(2).putShort(numberOfUncles).array()[1];
+
+        byte[] blockBeingMinedHeight = getBlockBeingMinedHeight(mainchainBlockHeaders);
+        System.arraycopy(blockBeingMinedHeight, 0, forkDetectionData, 8, 4);
+
+        return forkDetectionData;
+    }
+
+    private byte[] buildCommitToParentsVector(List<BlockHeader> mainchainBlocks) {
+        NetworkParameters params = RegTestParams.get();
```

Comment:
How are you sure that RegTestParams work for Bitcoin MainNet here ?

Reply:
Because here we are creating blocks in isolation and not running any validations on them. Furthermore, these blocks do not even have transactions. It's basically a creation of a block based on its hex header value.
I agree it would be a good idea to use the right network config and I'm creating an issue to change that on a next version.